nbclient/execute.py

```diff
@@ -572,8 +598,11 @@ def run_cell(self, cell, cell_index=0, store_history=False):
             except CellExecutionComplete:
                 more_output = False
 
-        # Return cell.outputs still for backwards compatibility
-        return exec_reply, cell.outputs
+        if execution_count:
+            cell['execution_count'] = execution_count
+        self._check_raise_for_error(cell, exec_reply)
+        self.nb['cells'][cell_index] = cell
```

Comment:
since we've already got access to the notebook in `self.nb`, couldn't this function be simplified by simply providing a `cell_index` rather than `cell` and `cell_index`? Then the first thing that could be done is running

```
cell = self.nb['cells'][cell_index] 
```

Reply:
> since we've already got access to the notebook in `self.nb`

I think that's actually a bad thing.
The notebook should *not* be stored in the executor!

This mistake was made in `nbconvert`, where the notebook was passed to the `preprocess()` method and subsequently stored in the executor:

https://github.com/jupyter/nbconvert/blob/8c9bffc8deb65ced99b20684ddd148fd885ad9e8/nbconvert/preprocessors/execute.py#L404

A later `preprocess()` call could overwrite the locally stored notebook.
Therefore, the `preprocess()` call was not re-entrant.

I've already mentioned this problem in `nbconvert`: https://github.com/jupyter/nbconvert/pull/886

During the move to `nbclient`, the situation seems to have gotten worse!

Now the notebook seems to be passed to the constructor of the executor, which means one executor instance can only ever execute a single notebook.
Is that really what you want?

To me, it would make much more sense if one executor instance could execute multiple notebooks in a row. Ideally also concurrently.

Reply:
> Now the notebook seems to be passed to the constructor of the executor, which means one executor instance can only ever execute a single notebook.
Is that really what you want?

Yes. There's a lot of state it holds for this particular execution and all the configuration for the execution means that having a owning object makes it a lot easier to manage this state and object manipulation. Otherwise we would have to pass many many arguments to every function, making the function signatures unwieldy.

> A later preprocess() call could overwrite the locally stored notebook.
Therefore, the preprocess() call was not re-entrant.
> ...
>
> During the move to nbclient, the situation seems to have gotten worse!

The issue in nbconvert was that it was half-way between abstractions. There were some functions that were reusable and some that were not without a new object. This class should now be re-entrant on execution methods at the top. `.execute` called twice will execute the whole notebook twice without having prior run state cause issues (I should add a test for this). I don't think it's "worse", it just more an object oriented approach than a functional one. in this case with lots of state and configuration object oriented has a lot of benefits over functional.

The rename that merged for changing to NotebookClient should help with the abstraction of concerns. Specifically it's denoting it acts on behalf of a notebook, meaning it should hold the state of the said notebook.

> To me, it would make much more sense if one executor instance could execute multiple notebooks in a row. Ideally also concurrently.

I didn't design for this pattern. We could have designed for that but I choose to keep it closer to how the code was before to reduce the number of changes -- also because I'm not sure that's strictly a better model. You do have the `execute` method for repeating calls to the client object to execute everything.

Reply:
> I didn't design for this pattern. We could have designed for that but I choose to keep it closer to how the code was before to reduce the number of changes -- also because I'm not sure that's strictly a better model.

This is pragmatic. I'd like to see the code evolve as well to be more functional, but at the same time we're still in this annoying mess of very class oriented dynamic traitlets for configuration in Jupyterland.